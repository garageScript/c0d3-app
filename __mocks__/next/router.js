const push = jest.fn()
const query = {}
export const useRouter = jest.fn().mockImplementation(() => {
  return {
    route: '/',
    pathname: '/',
    query,
    asPath: '/',
    push: push,
    replace: async () => true,
    reload: () => null,
    back: () => null
  }
})
