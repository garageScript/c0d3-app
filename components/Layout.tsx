import React from 'react'
import AppNav from './AppNav'
import Footer from './Footer'
import Head from 'next/head'
const Layout: React.FC<{ title?: string }> = ({ children, title }) => (
  <>
    <Head>
      <title key="title">{title ? `${title} â€” C0D3` : 'C0D3'}</title>
    </Head>
    <input type="text"></input>
    <AppNav />
    <div className="container-md">{children}</div>
    <Footer />
  </>
)

export default Layout
