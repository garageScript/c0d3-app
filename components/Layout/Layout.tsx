import React from 'react'
import AppNav from '../AppNav'
import Footer from '../Footer'
import Head from 'next/head'
import type { LayoutGetter } from '../../@types/page'
import styles from './layout.module.scss'

const Layout: React.FC<{ title?: string }> = ({ children, title }) => (
  <>
    <Head>
      <title key="title">{title ? `${title} â€” C0D3` : 'C0D3'}</title>
    </Head>
    <AppNav />
    <div className={`container-md ${styles.container}`}>{children}</div>
    <Footer />
  </>
)

export const getLayout: LayoutGetter = page => <Layout>{page}</Layout>
export default Layout
