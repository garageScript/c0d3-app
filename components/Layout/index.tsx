export { getLayout, default } from './Layout'
