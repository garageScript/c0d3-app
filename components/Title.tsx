import React from 'react'
import Head from 'next/head'

const Title: React.FC<{ title?: string }> = ({ title }) => (
  <Head>
    <title>{title ? `${title} ? C0D3` : 'C0D3'}</title>
  </Head>
)

export default Title
