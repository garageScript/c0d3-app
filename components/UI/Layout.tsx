import React from 'react'

const Layout: React.FC = ({ children }) => {
  return (
    <div className="container">
      {children}
      <style jsx>{`
        .container {
          display: grid;
          grid-template-columns: repeat(12, 1fr);
          grid-gap: 1em;
          max-width: 1440px;
          margin: auto;
          text-align: center;
          margin-bottom: 1em;
        }

        /* 
         * We can pre setup default column classes like below
         * Personally I prefear write this css into my components
         */
        .col-1 {
          grid-column: span 1;
          background-color: #e2e8f0;
        }
        .col-2 {
          grid-column: span 2;
          background-color: #e2e8f0;
        }
        .col-3 {
          grid-column: span 3;
          background-color: #e2e8f0;
        }
        .col-4 {
          grid-column: span 4;
          background-color: #e2e8f0;
        }
        .col-5 {
          grid-column: span 5;
          background-color: #e2e8f0;
        }
        .col-6 {
          grid-column: span 6;
          background-color: #e2e8f0;
        }
        .col-7 {
          grid-column: span 7;
          background-color: #e2e8f0;
        }
        .col-8 {
          grid-column: span 8;
          background-color: #e2e8f0;
        }
        .col-9 {
          grid-column: span 9;
          background-color: #e2e8f0;
        }
        .col-10 {
          grid-column: span 10;
          background-color: #e2e8f0;
        }
        .col-11 {
          grid-column: span 11;
          background-color: #e2e8f0;
        }
        .col-12 {
          grid-column: span 12;
          background-color: #e2e8f0;
        }
      `}</style>
    </div>
  )
}

export default Layout
