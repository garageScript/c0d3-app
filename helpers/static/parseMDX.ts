import { serialize } from 'next-mdx-remote/serialize'
import { MDXRemoteSerializeResult } from 'next-mdx-remote'
import matter from 'gray-matter'
import gfm from 'remark-gfm'
import slug from 'remark-slug'
import autolink from 'rehype-autolink-headings'

type ParsedMDX = {
  (
    fileContents: Buffer,
    onlyFront?: boolean
  ): Promise<{
    source?: MDXRemoteSerializeResult<Record<string, unknown>>
    frontMatter: { [key: string]: any }
  }>
}

const getSource = (content: string) =>
  serialize(content, {
    mdxOptions: {
      remarkPlugins: [slug, gfm],
      rehypePlugins: [autolink]
    }
  })

export const parseMDX: ParsedMDX = async (fileContents, onlyFront = false) => {
  const { content, data } = matter(fileContents)

  if (onlyFront) return { frontMatter: data }

  return {
    frontMatter: data,
    source: await getSource(content)
  }
}
