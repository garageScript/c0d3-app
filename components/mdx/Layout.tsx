import React from 'react'
import Layout from '../Layout'

const LessonLayout: React.FC = ({ children }) => {
  return (
    <Layout>
      <div className={`card shadow-sm mt-3 d-block border-0 p-4 bg-white`}>
        {children}
      </div>
    </Layout>
  )
}
export default LessonLayout
