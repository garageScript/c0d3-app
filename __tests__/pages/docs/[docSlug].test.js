import React from 'react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom'
import { dummyDocFileContent, dummyParsedDocMdx } from '../../../__dummy__/mdx'
import Docs, {
  getStaticProps,
  getStaticPaths
} from '../../../pages/docs/[docSlug]'
import { getLayout } from '../../../components/Layout'
import {
  getDocSlugs,
  getDocGithubFilePath,
  getDocContent
} from '../../../helpers/static/docs'
jest.mock('../../../helpers/static/docs')
import { parseMDX } from '../../../helpers/static/parseMDX'
jest.mock('../../../helpers/static/parseMDX')
import Title from '../../../components/Title'
jest.mock('../../../components/Title', () => {
  return jest.fn(() => null)
})

import { useRouter } from 'next/router' // Auto mocked

describe('[docSlug]', () => {
  const fakeGithubPath = 'some/fake/path'

  beforeEach(() => {
    jest.clearAllMocks()
  })
  describe('getStaticPaths', () => {
    test('should call getDocsSlugs helper function and return paths w/ fallback false', async () => {
      getDocSlugs.mockImplementation(() => [
        { docSlug: 'some_doc' },
        { docSlug: 'some_doc_two' }
      ])

      expect(await getStaticPaths()).toEqual({
        paths: [
          { params: { docSlug: 'some_doc' } },
          { params: { docSlug: 'some_doc_two' } }
        ],
        fallback: false
      })
    })
  })
  describe('getStaticProps', () => {
    test('should throw when called with no slug', async () => {
      await expect(() => getStaticProps({ params: {} })).rejects.toThrowError(
        /Missing Slug/
      )
    })

    test('should return correct props', async () => {
      getDocGithubFilePath.mockReturnValue(fakeGithubPath)
      getDocContent.mockReturnValue(dummyDocFileContent)
      parseMDX.mockResolvedValue(dummyParsedDocMdx)

      const docSlug = 'some_doc'
      expect(await getStaticProps({ params: { docSlug } })).toEqual({
        props: {
          ...dummyParsedDocMdx,
          docFilePath: fakeGithubPath
        }
      })
      expect(getDocGithubFilePath).toHaveBeenCalledWith(docSlug)
      expect(getDocContent).toHaveBeenCalledWith(docSlug)
    })
  })

  describe('Page', () => {
    test('Should use getLayout from Layout component', () => {
      expect(Docs.getLayout).toBe(getLayout)
    })

    test('should render functioning "Go Back" button', () => {
      const { push } = useRouter()
      render(<Docs {...dummyParsedDocMdx} docFilePath={fakeGithubPath} />)

      expect(push).not.toHaveBeenCalled()
      userEvent.click(screen.getByRole('button', { name: 'Go Back' }))
      expect(push).toHaveBeenCalledWith('/')
    })

    test('should render EditPage component with link to docFilePath', () => {
      render(<Docs {...dummyParsedDocMdx} docFilePath={fakeGithubPath} />)
      expect(screen.getByRole('link')).toHaveAttribute(
        'href',
        expect.stringContaining(fakeGithubPath)
      )
    })

    test('should display default document title when no frontMatter.title is provided', () => {
      render(
        <Docs
          frontMatter={{}}
          source={dummyParsedDocMdx.source}
          docFilePath={fakeGithubPath}
        />
      )

      expect(Title).toHaveBeenCalledWith({ title: 'C0D3 | Docs' }, {})
    })

    test('should render Title component with frontMatter.title', () => {
      render(<Docs {...dummyParsedDocMdx} docFilePath={fakeGithubPath} />)

      expect(Title).toHaveBeenCalledWith(
        {
          title: expect.stringContaining(dummyParsedDocMdx.frontMatter.title)
        },
        {}
      )
    })

    test('should render functioning MDX content', () => {
      render(<Docs {...dummyParsedDocMdx} docFilePath={fakeGithubPath} />)

      // Spoiler component renders into details / summary html elements
      const detailsElement = screen.getByRole('group')

      expect(detailsElement).not.toHaveAttribute('open')
      userEvent.click(screen.getByText('what am i?'))
      expect(detailsElement).toHaveAttribute('open')
    })
    test('should match screenshot', () => {
      const { container } = render(
        <Docs {...dummyParsedDocMdx} docFilePath={fakeGithubPath} />
      )

      expect(container).toMatchSnapshot()
    })
  })
})
