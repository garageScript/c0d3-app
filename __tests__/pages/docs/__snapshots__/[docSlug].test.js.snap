// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[docSlug] Page should match screenshot 1`] = `
<div>
  <div
    class="lesson-wrapper card shadow-sm mt-3 d-block border-0 p-3 p-md-4 bg-white"
  >
    <button
      class="link btn btn-link link-primary p-0"
    >
      Go Back
    </button>
    <div
      class="title"
    >
      fake doc
    </div>
    <h1
      class="MDX_h1"
      id="markdown"
    >
      markdown
    </h1>
    <details
      class="mb-3"
    >
      <summary>
        what am i?
      </summary>
      <div
        class="content"
      >
         mdx to jsx :)
      </div>
    </details>
    <a
      class="MDX_a edit"
      href="https://github.com/garageScript/c0d3-app/edit/master/some/fake/path"
      title="Edit this page on Github"
    >
      Edit this page on Github
    </a>
  </div>
</div>
`;
